From 71fd3a2420f6775feb11be05ef7e05895906c505 Mon Sep 17 00:00:00 2001
From: e2dk4r <43293320+e2dk4r@users.noreply.github.com>
Date: Sun, 7 Apr 2024 23:00:45 +0300
Subject: [PATCH] fix timeout

---
 src/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/main.c b/src/main.c
index cec9547..2754614 100644
--- a/src/main.c
+++ b/src/main.c
@@ -182,7 +182,7 @@ main(void)
 {
   int error_code = 0;
   struct wl_context context = {};
-  u64 timeout = 300;
+  u64 timeout = 30;
 
   /* wayland */
   context.wl_display = wl_display_connect(0);
-- 
2.43.2

