--- a/drivers/input/joystick/xpad.c        2024-06-16 01:25:16.016086953 +0300                                                                                  
+++ b/drivers/input/joystick/xpad.c        2024-06-16 01:25:06.264034605 +0300                                                                                  
@@ -381,6 +381,7 @@                                                                                                                                             
         { 0x31e3, 0x1310, "Wooting 60HE (ARM)", 0, XTYPE_XBOX360 },                                                                                            
         { 0x3285, 0x0607, "Nacon GC-100", 0, XTYPE_XBOX360 },                                                                                                  
         { 0x3537, 0x1004, "GameSir T4 Kaleid", 0, XTYPE_XBOX360 },                                                                                             
+        { 0x3537, 0x1010, "GameSir-G7 SE", 0, XTYPE_XBOXONE },                                                                                                 
         { 0x3767, 0x0101, "Fanatec Speedster 3 Forceshock Wheel", 0, XTYPE_XBOX },                                                                             
         { 0xffff, 0xffff, "Chinese-made Xbox Controller", 0, XTYPE_XBOX },                                                                                     
         { 0x0000, 0x0000, "Generic X-Box pad", 0, XTYPE_UNKNOWN }
